# Eksperimen Machine Learning - Heart Disease Dataset

**Nama:** ardir  
**Kelas:** Membangun Sistem Machine Learning  
**Dataset:** Heart Disease (Cleveland) - UCI ML Repository  
**Tanggal:** 2026-05-19

---

## Daftar Isi
1. Data Loading
2. Exploratory Data Analysis (EDA)
3. Data Preprocessing
4. Export Hasil Preprocessing
5. Modeling (Eksperimen Awal)
6. Evaluasi Model
7. Kesimpulan

## 1. Data Loading

Pada tahap ini, kita akan memuat dataset Heart Disease (Cleveland) dari UCI ML Repository. Dataset ini berisi data klinis pasien untuk prediksi penyakit jantung.

In [1]:
# Import library yang diperlukan
import os
import pandas as pd
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score, f1_score,
    classification_report, confusion_matrix
)
import joblib
import warnings
warnings.filterwarnings("ignore")

print("Library berhasil dimuat!")

Library berhasil dimuat!


In [2]:
# Load dataset dari UCI ML Repository
UCI_URL = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data"

column_names = [
    "age", "sex", "cp", "trestbps", "chol", "fbs",
    "restecg", "thalach", "exang", "oldpeak", "slope",
    "ca", "thal", "target"
]

df = pd.read_csv(UCI_URL, names=column_names, na_values="?")
df["target"] = (df["target"] > 0).astype(int)

print(f"Total Dataset: {len(df)} baris, {df.shape[1]} kolom")
print(f"No Disease: {(df['target']==0).sum()}, Disease: {(df['target']==1).sum()}")

Total Dataset: 303 baris, 14 kolom
No Disease: 164, Disease: 139


In [3]:
df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63.0,1.0,1.0,145.0,233.0,1.0,2.0,150.0,0.0,2.3,3.0,0.0,6.0,0
1,67.0,1.0,4.0,160.0,286.0,0.0,2.0,108.0,1.0,1.5,2.0,3.0,3.0,1
2,67.0,1.0,4.0,120.0,229.0,0.0,2.0,129.0,1.0,2.6,2.0,2.0,7.0,1
3,37.0,1.0,3.0,130.0,250.0,0.0,0.0,187.0,0.0,3.5,3.0,0.0,3.0,0
4,41.0,0.0,2.0,130.0,204.0,0.0,2.0,172.0,0.0,1.4,1.0,0.0,3.0,0


In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    float64
 1   sex       303 non-null    float64
 2   cp        303 non-null    float64
 3   trestbps  303 non-null    float64
 4   chol      303 non-null    float64
 5   fbs       303 non-null    float64
 6   restecg   303 non-null    float64
 7   thalach   303 non-null    float64
 8   exang     303 non-null    float64
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    float64
 11  ca        299 non-null    float64
 12  thal      301 non-null    float64
 13  target    303 non-null    int64  
dtypes: float64(13), int64(1)
memory usage: 33.3 KB


In [5]:
df.describe()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
count,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,299.000000,301.000000,303.000000
mean,54.438944,0.679868,3.158416,131.689769,246.693069,0.148515,0.990099,149.607261,0.326733,1.039604,1.600660,0.672241,4.734219,0.458746
std,9.038662,0.467299,0.960126,17.599748,51.776918,0.356198,0.994971,22.875003,0.469794,1.161075,0.616226,0.937438,1.939706,0.499120
min,29.000000,0.000000,1.000000,94.000000,126.000000,0.000000,0.000000,71.000000,0.000000,0.000000,1.000000,0.000000,3.000000,0.000000
25%,48.000000,0.000000,3.000000,120.000000,211.000000,0.000000,0.000000,133.500000,0.000000,0.000000,1.000000,0.000000,3.000000,0.000000
50%,56.000000,1.000000,3.000000,130.000000,241.000000,0.000000,1.000000,153.000000,0.000000,0.800000,2.000000,0.000000,3.000000,0.000000
75%,61.000000,1.000000,4.000000,140.000000,275.000000,0.000000,2.000000,166.000000,1.000000,1.600000,2.000000,1.000000,7.000000,1.000000
max,77.000000,1.000000,4.000000,200.000000,564.000000,1.000000,2.000000,202.000000,1.000000,6.200000,3.000000,3.000000,7.000000,1.000000


## 2. Exploratory Data Analysis (EDA)

Pada tahap ini kita akan mengeksplorasi distribusi data, korelasi antar fitur, dan keseimbangan kelas.

In [6]:
# Cek missing values
print("Missing Values:")
print(df.isnull().sum())
print(f"\nTotal missing values: {df.isnull().sum().sum()}")

Missing Values:
age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        2
target      0
dtype: int64

Total missing values: 6


In [7]:
print(f"Jumlah baris duplikat: {df.duplicated().sum()}")

Jumlah baris duplikat: 0


In [8]:
# Distribusi target
fig, axes = plt.subplots(1, 2, figsize=(14, 5))
target_counts = df["target"].value_counts().sort_index()
colors = ["#2196F3", "#FF5722"]
axes[0].bar(["No Disease (0)", "Disease (1)"], target_counts.values, color=colors)
axes[0].set_xlabel("Target")
axes[0].set_ylabel("Count")
axes[0].set_title("Distribusi Target Heart Disease")
axes[1].pie(target_counts, labels=["No Disease", "Disease"], autopct="%1.1f%%",
            colors=colors, startangle=90)
axes[1].set_title("Proporsi No Disease vs Disease")
plt.tight_layout()
plt.savefig('eda_target.png', dpi=100)
plt.show()

In [9]:
# Distribusi fitur numerik
num_cols = df.select_dtypes(include=[np.number]).columns.tolist()
fig, axes = plt.subplots(4, 4, figsize=(16, 14))
axes = axes.flatten()
for i, col in enumerate(num_cols):
    if i < len(axes):
        axes[i].hist(df[col].dropna(), bins=30, color="#3498DB", alpha=0.7, edgecolor="black")
        axes[i].set_title(col, fontsize=10)
        axes[i].set_ylabel("Frequency")
for j in range(len(num_cols), len(axes)):
    fig.delaxes(axes[j])
plt.suptitle("Distribusi Fitur Numerik", fontsize=14, fontweight="bold")
plt.tight_layout()
plt.savefig('eda_distributions.png', dpi=100)
plt.show()

In [10]:
# Correlation heatmap
plt.figure(figsize=(12, 10))
corr_matrix = df[num_cols].corr()
mask = np.triu(np.ones_like(corr_matrix, dtype=bool))
sns.heatmap(corr_matrix, mask=mask, annot=True, fmt=".2f",
            cmap="coolwarm", center=0, linewidths=0.5,
            square=True, cbar_kws={"shrink": 0.8})
plt.title("Korelasi Antar Fitur", fontsize=14, fontweight="bold")
plt.tight_layout()
plt.savefig('eda_correlation.png', dpi=100)
plt.show()

In [11]:
# Boxplot per target
fig, axes = plt.subplots(2, 3, figsize=(16, 10))
axes = axes.flatten()
important_features = ["age", "trestbps", "chol", "thalach", "oldpeak", "ca"]
for i, col in enumerate(important_features):
    sns.boxplot(data=df, x="target", y=col, ax=axes[i], palette="viridis")
    axes[i].set_title(f"{col} vs Target")
plt.suptitle("Fitur Penting vs Target", fontsize=14, fontweight="bold")
plt.tight_layout()
plt.savefig('eda_boxplots.png', dpi=100)
plt.show()

## 3. Data Preprocessing

Tahapan preprocessing meliputi:
- Handling missing values & duplicates
- Feature engineering
- Scaling fitur numerik
- Train-validation-test split

In [12]:
# 3.1 Handle missing values & duplikasi
print(f"Sebelum cleaning: {len(df)} baris")
df_clean = df.copy()

# Isi missing values dengan median untuk semua kolom
for col in df_clean.columns:
    if df_clean[col].isnull().sum() > 0:
        median_val = df_clean[col].median()
        df_clean[col] = df_clean[col].fillna(median_val)
        print(f"  {col}: {df[col].isnull().sum()} missing -> filled with median ({median_val})")

# Hapus duplikasi
df_clean = df_clean.drop_duplicates()
print(f"Sesudah cleaning: {len(df_clean)} baris")
print(f"Remaining NaN: {df_clean.isnull().sum().sum()}")

Sebelum cleaning: 303 baris
  ca: 4 missing -> filled with median (0.0)
  thal: 2 missing -> filled with median (3.0)
Sesudah cleaning: 303 baris
Remaining NaN: 0


In [13]:
# 3.2 Feature Engineering
df_eng = df_clean.copy()
df_eng["heart_rate_reserve"] = df_eng["thalach"] / np.maximum(220 - df_eng["age"], 1)
df_eng["cholesterol_age_ratio"] = df_eng["chol"] / np.maximum(df_eng["age"], 1)
df_eng["bp_chol_interaction"] = df_eng["trestbps"] * df_eng["chol"] / 10000
df_eng["exercise_risk"] = df_eng["oldpeak"] * (df_eng["exang"] + 1)
df_eng["age_sex_risk"] = df_eng["age"] * (1 + df_eng["sex"] * 0.2)

print(f"Total fitur setelah engineering: {df_eng.shape[1]}")
print(f"Distribusi target:")
print(df_eng["target"].value_counts())
print(f"\nVerifikasi NaN setelah engineering: {df_eng.isnull().sum().sum()}")

Total fitur setelah engineering: 19
Distribusi target:
target
0    164
1    139
Name: count, dtype: int64

Verifikasi NaN setelah engineering: 0


In [14]:
# 3.3 Definisikan fitur dan target
X = df_eng.drop(columns=["target"])
y = df_eng["target"]
print(f"Fitur shape: {X.shape}")
print(f"Target shape: {y.shape}")
print(f"Nama fitur: {X.columns.tolist()}")

Fitur shape: (303, 18)
Target shape: (303,)
Nama fitur: ['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'heart_rate_reserve', 'cholesterol_age_ratio', 'bp_chol_interaction', 'exercise_risk', 'age_sex_risk']


In [15]:
# 3.4 Train-Validation-Test Split
X_temp, X_test, y_temp, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)
X_train, X_val, y_train, y_val = train_test_split(X_temp, y_temp, test_size=0.125, random_state=42, stratify=y_temp)
total = len(X)
print(f"Train: {X_train.shape[0]} samples ({X_train.shape[0]/total*100:.1f}%)")
print(f"Val:   {X_val.shape[0]} samples ({X_val.shape[0]/total*100:.1f}%)")
print(f"Test:  {X_test.shape[0]} samples ({X_test.shape[0]/total*100:.1f}%)")

Train: 211 samples (69.6%)
Val:   31 samples (10.2%)
Test:  61 samples (20.1%)


In [16]:
# 3.5 Scaling
scaler = StandardScaler()
num_cols_scale = X_train.select_dtypes(include=[np.number]).columns.tolist()
X_train_scaled = X_train.copy()
X_val_scaled = X_val.copy()
X_test_scaled = X_test.copy()
X_train_scaled[num_cols_scale] = scaler.fit_transform(X_train[num_cols_scale])
X_val_scaled[num_cols_scale] = scaler.transform(X_val[num_cols_scale])
X_test_scaled[num_cols_scale] = scaler.transform(X_test[num_cols_scale])

# Final NaN check
assert X_train_scaled.isnull().sum().sum() == 0, "Train data still has NaN!"
assert X_val_scaled.isnull().sum().sum() == 0, "Val data still has NaN!"
assert X_test_scaled.isnull().sum().sum() == 0, "Test data still has NaN!"
print("Scaling selesai! Tidak ada NaN.")
X_train_scaled.head()

Scaling selesai! Tidak ada NaN.


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,heart_rate_reserve,cholesterol_age_ratio,bp_chol_interaction,exercise_risk,age_sex_risk
151,-1.383388,-1.419255,0.877112,-1.642313,0.290100,-0.430573,1.061111,-1.213205,-0.674665,-0.366659,0.717656,-0.703350,-0.882785,-1.721918,1.485055,-0.665344,-0.472177,-1.790113
145,-0.823988,0.704595,-0.186512,-1.298535,-0.128502,-0.430573,-0.946911,0.123305,-0.674665,-0.903296,-0.928272,-0.703350,-0.882785,-0.176517,0.440256,-0.756718,-0.789131,-0.472570
255,-1.383388,-1.419255,-0.186512,-0.610979,-0.775433,-0.430573,-0.946911,1.058862,-0.674665,-0.903296,0.717656,-0.703350,-0.882785,0.569691,0.262329,-0.892035,-0.789131,-1.790113
111,0.182932,0.704595,0.877112,-0.324498,-0.014338,2.322486,1.061111,-0.233097,1.482218,0.169977,0.717656,0.438241,-0.882785,-0.181027,-0.223487,-0.189293,0.478686,0.515587
70,1.189852,-1.419255,-0.186512,1.394391,0.366209,-0.430573,-0.946911,-0.054896,-0.674665,-0.187781,-0.928272,-0.703350,-0.882785,0.433151,-0.505907,1.039488,-0.366525,0.314295


## 4. Export Hasil Preprocessing

Menyimpan hasil preprocessing ke file CSV dan pkl agar dapat digunakan di tahap selanjutnya (model building, CI workflow, dan monitoring).

In [17]:
# 4.1 Buat direktori output
output_dir = 'heart_disease_preprocessing'
os.makedirs(output_dir, exist_ok=True)
print(f"Direktori output: {output_dir}/")

Direktori output: heart_disease_preprocessing/


In [18]:
# 4.2 Simpan dataset train, validation, dan test ke CSV
X_train_scaled.to_csv(os.path.join(output_dir, 'X_train.csv'), index=False)
X_val_scaled.to_csv(os.path.join(output_dir, 'X_val.csv'), index=False)
X_test_scaled.to_csv(os.path.join(output_dir, 'X_test.csv'), index=False)
y_train.to_csv(os.path.join(output_dir, 'y_train.csv'), index=False)
y_val.to_csv(os.path.join(output_dir, 'y_val.csv'), index=False)
y_test.to_csv(os.path.join(output_dir, 'y_test.csv'), index=False)

print("Dataset berhasil disimpan:")
print(f"  X_train.csv: {X_train_scaled.shape}")
print(f"  X_val.csv:   {X_val_scaled.shape}")
print(f"  X_test.csv:  {X_test_scaled.shape}")
print(f"  y_train.csv: {y_train.shape}")
print(f"  y_val.csv:   {y_val.shape}")
print(f"  y_test.csv:  {y_test.shape}")

Dataset berhasil disimpan:
  X_train.csv: (211, 18)
  X_val.csv:   (31, 18)
  X_test.csv:  (61, 18)
  y_train.csv: (211,)
  y_val.csv:   (31,)
  y_test.csv:  (61,)


In [19]:
# 4.3 Simpan scaler dan encoder ke file pkl
joblib.dump(scaler, os.path.join(output_dir, 'scaler.pkl'))
print(f"Scaler disimpan: {output_dir}/scaler.pkl")

# Simpan label encoder untuk target
le_target = LabelEncoder()
le_target.classes_ = np.array(['no_disease', 'disease'])
encoders = {'target': le_target}
joblib.dump(encoders, os.path.join(output_dir, 'encoders.pkl'))
print(f"Encoders disimpan: {output_dir}/encoders.pkl")

# Simpan feature names
pd.Series(num_cols_scale).to_csv(os.path.join(output_dir, 'feature_names.csv'), index=False)
print(f"Feature names disimpan: {output_dir}/feature_names.csv")

Scaler disimpan: heart_disease_preprocessing/scaler.pkl
Encoders disimpan: heart_disease_preprocessing/encoders.pkl
Feature names disimpan: heart_disease_preprocessing/feature_names.csv


In [20]:
# 4.4 Verifikasi file yang telah disimpan
print("\nFile yang berhasil di-export:")
for f in sorted(os.listdir(output_dir)):
    size = os.path.getsize(os.path.join(output_dir, f))
    print(f"  {f}: {size:,} bytes")
print(f"\nTotal: {len(os.listdir(output_dir))} files")


File yang berhasil di-export:
  X_test.csv: 21,774 bytes
  X_train.csv: 74,986 bytes
  X_val.csv: 11,114 bytes
  encoders.pkl: 424 bytes
  feature_names.csv: 182 bytes
  scaler.pkl: 1,463 bytes
  y_test.csv: 191 bytes
  y_train.csv: 641 bytes
  y_val.csv: 101 bytes

Total: 9 files


## 5. Modeling (Eksperimen Awal)

Kita akan mencoba beberapa model:
1. Random Forest Classifier
2. Gradient Boosting Classifier
3. Logistic Regression

In [21]:
# 5.1 Random Forest
rf_model = RandomForestClassifier(n_estimators=100, max_depth=10, random_state=42, n_jobs=-1)
rf_model.fit(X_train_scaled, y_train)
rf_train_acc = accuracy_score(y_train, rf_model.predict(X_train_scaled))
rf_val_acc = accuracy_score(y_val, rf_model.predict(X_val_scaled))
rf_test_acc = accuracy_score(y_test, rf_model.predict(X_test_scaled))
print("Random Forest Results:")
print(f"  Train Accuracy: {rf_train_acc:.4f}")
print(f"  Val Accuracy:   {rf_val_acc:.4f}")
print(f"  Test Accuracy:  {rf_test_acc:.4f}")

Random Forest Results:
  Train Accuracy: 1.0000
  Val Accuracy:   0.8065
  Test Accuracy:  0.9344


In [22]:
# 5.2 Gradient Boosting
gb_model = GradientBoostingClassifier(n_estimators=100, max_depth=5, learning_rate=0.1, random_state=42)
gb_model.fit(X_train_scaled, y_train)
gb_train_acc = accuracy_score(y_train, gb_model.predict(X_train_scaled))
gb_val_acc = accuracy_score(y_val, gb_model.predict(X_val_scaled))
gb_test_acc = accuracy_score(y_test, gb_model.predict(X_test_scaled))
print("Gradient Boosting Results:")
print(f"  Train Accuracy: {gb_train_acc:.4f}")
print(f"  Val Accuracy:   {gb_val_acc:.4f}")
print(f"  Test Accuracy:  {gb_test_acc:.4f}")

Gradient Boosting Results:
  Train Accuracy: 1.0000
  Val Accuracy:   0.8387
  Test Accuracy:  0.8852


In [23]:
# 5.3 Logistic Regression
lr_model = LogisticRegression(max_iter=1000, random_state=42)
lr_model.fit(X_train_scaled, y_train)
lr_train_acc = accuracy_score(y_train, lr_model.predict(X_train_scaled))
lr_val_acc = accuracy_score(y_val, lr_model.predict(X_val_scaled))
lr_test_acc = accuracy_score(y_test, lr_model.predict(X_test_scaled))
print("Logistic Regression Results:")
print(f"  Train Accuracy: {lr_train_acc:.4f}")
print(f"  Val Accuracy:   {lr_val_acc:.4f}")
print(f"  Test Accuracy:  {lr_test_acc:.4f}")

Logistic Regression Results:
  Train Accuracy: 0.8483
  Val Accuracy:   0.8387
  Test Accuracy:  0.8852


## 6. Evaluasi Model

Membandingkan performa semua model dan mengevaluasi model terbaik secara detail.

In [24]:
# 6.1 Perbandingan Model
results = pd.DataFrame({
    "Model": ["Random Forest", "Gradient Boosting", "Logistic Regression"],
    "Train Acc": [rf_train_acc, gb_train_acc, lr_train_acc],
    "Val Acc": [rf_val_acc, gb_val_acc, lr_val_acc],
    "Test Acc": [rf_test_acc, gb_test_acc, lr_test_acc]
})
print("Perbandingan Performa Model:")
print(results.to_string(index=False))

fig, ax = plt.subplots(figsize=(10, 6))
x = np.arange(len(results))
width = 0.25
bars1 = ax.bar(x - width, results["Train Acc"], width, label="Train", color="#2196F3")
bars2 = ax.bar(x, results["Val Acc"], width, label="Validation", color="#4CAF50")
bars3 = ax.bar(x + width, results["Test Acc"], width, label="Test", color="#FF5722")
ax.set_ylabel("Accuracy")
ax.set_title("Perbandingan Performa Model", fontsize=14, fontweight="bold")
ax.set_xticks(x)
ax.set_xticklabels(results["Model"])
ax.legend()
ax.set_ylim(0, 1.1)
for bars in [bars1, bars2, bars3]:
    for bar in bars:
        height = bar.get_height()
        ax.annotate(f"{height:.3f}", xy=(bar.get_x() + bar.get_width()/2, height),
                    xytext=(0, 3), textcoords="offset points", ha="center", fontsize=8)
plt.tight_layout()
plt.savefig('eval_comparison.png', dpi=100)
plt.show()

Perbandingan Performa Model:
              Model  Train Acc  Val Acc  Test Acc
      Random Forest   1.000000 0.806452  0.934426
  Gradient Boosting   1.000000 0.838710  0.885246
Logistic Regression   0.848341 0.838710  0.885246


In [25]:
# 6.2 Classification Report
y_pred_test = rf_model.predict(X_test_scaled)
print("Classification Report - Random Forest (Test Set):")
print(classification_report(y_test, y_pred_test, target_names=["no_disease", "disease"]))

Classification Report - Random Forest (Test Set):


              precision    recall  f1-score   support

  no_disease       0.97      0.91      0.94        33
     disease       0.90      0.96      0.93        28

    accuracy                           0.93        61
   macro avg       0.93      0.94      0.93        61
weighted avg       0.94      0.93      0.93        61



In [26]:
# 6.3 Confusion Matrix
cm = confusion_matrix(y_test, y_pred_test)
plt.figure(figsize=(8, 6))
sns.heatmap(cm, annot=True, fmt="d", cmap="Blues",
            xticklabels=["no_disease", "disease"], yticklabels=["no_disease", "disease"])
plt.title("Confusion Matrix - Random Forest", fontsize=14, fontweight="bold")
plt.xlabel("Predicted")
plt.ylabel("Actual")
plt.tight_layout()
plt.savefig('eval_confusion_matrix.png', dpi=100)
plt.show()

In [27]:
# 6.4 Feature Importance
importances = rf_model.feature_importances_
indices = np.argsort(importances)[::-1]
plt.figure(figsize=(12, 6))
colors = plt.cm.viridis(np.linspace(0.3, 0.9, len(indices)))
plt.barh(range(len(indices)), importances[indices][::-1], color=colors)
plt.yticks(range(len(indices)), [X_train_scaled.columns[i] for i in indices[::-1]])
plt.xlabel("Importance")
plt.title("Feature Importances - Random Forest", fontsize=14, fontweight="bold")
plt.tight_layout()
plt.savefig('eval_feature_importance.png', dpi=100)
plt.show()

## 7. Kesimpulan

### Temuan Utama:
1. **Dataset**: Heart Disease (Cleveland) dataset terdiri dari 303 sampel dengan 13 fitur klinis dan 1 target.
2. **EDA**: Distribusi target cukup seimbang (164 no disease vs 139 disease). Fitur `cp`, `thalach`, `ca`, dan `oldpeak` memiliki korelasi kuat dengan target.
3. **Preprocessing**: Missing values pada kolom `ca` dan `thal` diisi dengan median. Feature engineering menambah 5 fitur baru.
4. **Export**: Seluruh hasil preprocessing (CSV, scaler, encoders) berhasil disimpan ke folder `heart_disease_preprocessing/`.
5. **Model Terbaik**: Random Forest Classifier memberikan performa terbaik di antara 3 model yang diuji.

### Langkah Selanjutnya:
- Implementasi automated preprocessing dengan `automate_ardir.py`
- Integrasi dengan MLflow untuk tracking eksperimen
- Hyperparameter tuning lebih lanjut
- Deployment model menggunakan FastAPI
- Monitoring dengan Prometheus dan Grafana